### ISSUE TITLE
insertOrUpdate compiles wrong statement on the table without soft syms.

### ISSUE THREAD

**Issue by janghwan** [slick/slick#1627]
insertOrUpdate compiles incorrect SQL statement when there is no other columns than primary key in the table.

I have a table which has only primary keys

``` sql
CREATE TABLE user_address (
  user_id INT,
  address_id INT,
  PRIMARY KEY (user_id, address_id)
);
```

``` scala
  class UserAddress(tag: Tag) extends Table[(Int, Int)](tag, "user_address") {
    def userId = column[Int]("user_id", O.PrimaryKey)
    def addressId = column[Int]("address_id", O.PrimaryKey)
    def * = (userId, addressId)
  }

  val userAddresses = TableQuery[UserAddress]
```

if i run this

``` scala
userAddresses.insertOrUpdate((uid, aid))
```

this generates following statement

``` sql
begin; update "user_address" set  where "user_id"=? and "addresss_id"=?; insert into "user_address" ("user_id","address_id") select ?,? where not exists (select 1 from "user_address" where "user_id"=? and "address_id"=?); end
```

,and it gives syntax error
`org.postgresql.util.PSQLException: ERROR: syntax error at or near "where"`

This is due to this line in `UpdateInsertBuilder`

``` scala
"update " + tableName + " set " + softNames.map(n => s"$n=?").mkString(",") + " where " + pkNames.map(n => s"$n=?").mkString(" and "),
```

If there is no column other than primary keys `set` clause will be empty. Is it intended or something need to be fixed?

I think `INSERT IGNORE` query would be more fit in this case, but I used insertOrUpdate since slick doesn't support it.

**Comment by tminglei** [comment]
@janghwan this is a bug. But I'm not sure how it will be fixed.
There's a related issue #1631, which wants to replace it with pg native upsert.

_You can watch it._

**Comment by hvesalai** [comment]
Duplicate of #1699

**Comment by SiddharthaDeshpande** [comment]
Hey,
As far as I followed the linked threads, the fix will be done in 3.2.3? or is it released in 3.2.2?

**Comment by hvesalai** [comment]
In 3.2.2 we have merged this pull request: https://github.com/slick/slick/pull/1836

**Comment by hvesalai** [comment]
Let's re-open this issue if the problem persists.

**Comment by SiddharthaDeshpande** [comment]
Hey yes, I did update my slick version to 3.2.2 but it seems code still looks same.

""  
      "update " + tableName + " set " + softNames.map(n => s"$n=?").mkString(",") + " where " + pkNames.map(n => s"$n=?").mkString(" and "),
        ConstArray.from(softSyms ++ pkSyms))
""
The corner case is when you have only columns which are marked as Primary key

**Comment by hvesalai** [comment]
Was there a PR to fix this?

**Comment by hvesalai** [comment]
And do we really want to support this, or should it just throw an Exception in this case also. After all, it will not do an insertOrUpdate, it will be -- as pointed out earlier -- insert ignore.

**Comment by SiddharthaDeshpande** [comment]
Hey, @hvesalai .
Absolutely. insertIgnore will be the right candidate.
I suppose insertIgnore will be supported in upcoming releases?

**Comment by hvesalai** [comment]
Slick is now fully a community project, so we welcome pull requests for both bug fixes and enhancements. So we need to wait until somebody comes and champions such a feature.

**Comment by hvesalai** [comment]
We need to discuss reverting the fix that solved this problem. The solution to this problem was to use "insert ignore", but that is now causing an other issue, namely #2076 

As I already anticipated in the above discussion, it might have just been the correct thing to do to have the exception thrown when a duplicate row was being inserted and have the program logic ignore that exception.

**Comment by ctataryn** [comment]
We definitely need to be able to insert on a table that contains only primary keys. (i.e. join tables). So if the original fix is to be reverted, then an alternate fix must be proposed.

 I don’t really understand why you’d want to ignore an attempt to insert duplicate rows though? Maybe I’m reading that wrong?

**Comment by hvesalai** [comment]
This issue is about "insert or ignore" to a table where there are only primary keys. The question is, do we really need such a primitive, because it can also be implemented with normal insert and then just catching the exception.

If you do that in an auto-commit session, then you can continue normally after catching the exception. However, if you are using transactions and want the statements before and after the exception to success, then it's sligtly more complex: you need to use a savepoint before the insert and revert back to that savepoint if the insert fails (after which you can continue normally and eventually commit the tx).

**Comment by nafg** [comment]
Closed by #2201
